File: python-algo/leetcode/problem_1335.py
"""1335. Minimum Difficulty of a Job Schedule
https://leetcode.com/problems/minimum-difficulty-of-a-job-schedule/
"""
from typing import List


class Solution:
    def min_difficulty(self, job_difficulty: List[int], d: int) -> int:
        n = len(job_difficulty)
        if d > n:
            return -1
        max_d = [[0] * n for _ in range(n)]  # max difficulty between [i:j]
        for i in range(n):
            max_d[i][i] = job_difficulty[i]
            for j in range(i + 1, n):
                max_d[i][j] = max(max_d[i][j - 1], job_difficulty[j])
        dp = [[-1] * (d + 1) for _ in range(n)]
        for i in range(n):
            dp[i][1] = max_d[0][i]
            for j in range(2, min(i + 1, d) + 1):
                dp[i][j] = 900000
                for k in range(i):
                    if dp[k][j - 1] > -1:
                        dp[i][j] = min(dp[i][j], dp[k][j - 1] + max_d[k + 1][i])
        return dp[-1][-1]

File: java-algo/src/main/java/com/leetcode/Problem1335.java
package com.leetcode;

import java.util.Arrays;

/**
 * 1335. Minimum Difficulty of a Job Schedule
 * https://leetcode.com/problems/minimum-difficulty-of-a-job-schedule/
 */
public class Problem1335 {
    public int minDifficulty(int[] jobDifficulty, int d) {
        int n = jobDifficulty.length;
        if (n < d) {
            return -1;
        }
        int[][] maxDifficulty = new int[n][n]; // 预计算从第 i 到第 j 项工作的最大难度
        for (int i = 0; i < n; i++) {
            maxDifficulty[i][i] = jobDifficulty[i];
            for (int j = i + 1; j < n; j++) {
                maxDifficulty[i][j] = Math.max(maxDifficulty[i][j - 1], jobDifficulty[j]);
            }
        }
        int[][] dp = new int[n][d + 1]; // dp[i][j] 表示截止到第 i 项工作，耗时 j 天的总难度
        for (int i = 0; i < n; i++) {
            Arrays.fill(dp[i], -1);
        }
        for (int i = 0; i < n; i++) {
            dp[i][1] = maxDifficulty[0][i];
            for (int j = 2; j <= Math.min(d, i + 1); j++) {
                dp[i][j] = Integer.MAX_VALUE;
                for (int k = 0; k < i; k++) {
                    if (dp[k][j - 1] > -1) {
                        // dp[i][j] = dp[k][j - 1] + max_difficulty[k+1][j]
                        dp[i][j] = Math.min(dp[i][j], dp[k][j - 1] + maxDifficulty[k + 1][i]);
                    }
                }
            }
        }
        return dp[n - 1][d];
    }
}
